xla/hlo/transforms/collectives/rematerialize_large_all_gather.cc

```diff
@@ -0,0 +1,165 @@
+      tp_replica_group_found = true;
+      tp_replica_group = rval.second;
+    }
+  }
+
+  for (HloComputation* computation : module->computations(execution_threads)) {
+    for (HloInstruction* inst : computation->instructions()) {
+      if (inst->opcode() == HloOpcode::kGetTupleElement) {
+        HloInstruction* opt_barrier = inst->mutable_operand(0);
+        if (opt_barrier->opcode() == HloOpcode::kOptimizationBarrier) {
+          int64_t tuple_index = inst->tuple_index();
+          HloInstruction* input_inst =
+              opt_barrier->mutable_operand(0)->mutable_operand(tuple_index);
+          if (input_inst->opcode() == HloOpcode::kAllGather &&
+              IsRemattableAllGather(tp_replica_group, input_inst)) {
+            VLOG(2) << "Will remat all-gather: " << input_inst->ToString();
+            // Replace the operand of the optimization barrier's tuple with the
+            // AllGather input
+            TF_RETURN_IF_ERROR(
+                opt_barrier->mutable_operand(0)
+                    ->ReplaceOperandWithDifferentShape(
+                        tuple_index, input_inst->mutable_operand(0)));
+            // Update the shape of the optimization barrier's tuple to match the
+            // new operand shape
+            Shape new_tuple_shape = opt_barrier->operand(0)->shape();
+            *new_tuple_shape.mutable_tuple_shapes(tuple_index) =
+                input_inst->mutable_operand(0)->shape();
+            *opt_barrier->mutable_operand(0)->mutable_shape() = new_tuple_shape;
+            // Update the shape of the optimization barrier to match the new
+            // tuple shape
+            *opt_barrier->mutable_shape() = new_tuple_shape;
+            // Replace the GetTupleElement with a new GetTupleElement from the
+            // updated OptimizationBarrier
+            HloInstruction* new_gte = computation->AddInstruction(
+                HloInstruction::CreateGetTupleElement(
+                    input_inst->mutable_operand(0)->shape(), opt_barrier,
+                    tuple_index));
+
+            HloAllGatherInstruction* all_gather =
+                Cast<HloAllGatherInstruction>(input_inst);
+            // Launch the AllGather on the new GetTupleElement
+            HloInstruction* new_allgather =
+                computation->AddInstruction(HloInstruction::CreateAllGather(
+                    all_gather->shape(), {new_gte},
+                    all_gather->all_gather_dimension(),
+                    all_gather->replica_groups(),
+                    all_gather->constrain_layout(), all_gather->channel_id(),
+                    all_gather->use_global_device_ids()));
+            // Replace the original GetTupleElement with the new AllGather
+            TF_RETURN_IF_ERROR(inst->ReplaceAllUsesWith(new_allgather));
+
+            changed = true;
+          }
+        }
+      }
+    }
+  }
+  if (changed) {
+    // Run DCE if changed.
+    HloDCE dce;
+    TF_RETURN_IF_ERROR(dce.Run(module).status());
```

Comment:
I dont think you want to run DCE in the pass itself but rather after the pass in the pipeline surrounding this

Reply:
Yes, I run DCE to remove the dead all-gathers.
What is recommended way to handle dead instructions after a pass? Should I manually call TF_REMOVE_FROM_COMPUTATION(instruction) myself?

Reply:
Done.